agent/immortalstreams/backedpipe/backed_pipe.go

```diff
@@ -0,0 +1,357 @@
+func (bp *BackedPipe) Read(p []byte) (int, error) {
+	return bp.reader.Read(p)
+}
+
+// Write implements io.Writer by delegating to the BackedWriter.
+func (bp *BackedPipe) Write(p []byte) (int, error) {
+	bp.mu.RLock()
+	writer := bp.writer
+	state := bp.state
+	bp.mu.RUnlock()
+
+	if state == closed {
+		return 0, io.EOF
+	}
+
+	return writer.Write(p)
+}
+
+// Close closes the pipe and all underlying connections.
+func (bp *BackedPipe) Close() error {
+	bp.mu.Lock()
+	defer bp.mu.Unlock()
+
+	if bp.state == closed {
+		return nil
+	}
+
+	bp.state = closed
+	bp.cancel() // Cancel main context
+
+	// Close all components in parallel to avoid deadlocks
+	//
+	// IMPORTANT: The connection must be closed first to unblock any
+	// readers or writers that might be holding the mutex on Read/Write
+	var g errgroup.Group
+
+	if bp.conn != nil {
+		conn := bp.conn
+		g.Go(func() error {
+			return conn.Close()
+		})
+		bp.conn = nil
+	}
+
+	if bp.reader != nil {
+		reader := bp.reader
+		g.Go(func() error {
+			return reader.Close()
+		})
+	}
+
+	if bp.writer != nil {
+		writer := bp.writer
+		g.Go(func() error {
+			return writer.Close()
+		})
+	}
+
+	// Wait for all close operations to complete and return any error
+	return g.Wait()
+}
+
+// Connected returns whether the pipe is currently connected.
+func (bp *BackedPipe) Connected() bool {
+	bp.mu.RLock()
+	defer bp.mu.RUnlock()
+	return bp.state == connected
+}
+
+// reconnectLocked handles the reconnection logic. Must be called with write lock held.
+func (bp *BackedPipe) reconnectLocked() error {
+	if bp.state == reconnecting {
+		return ErrReconnectionInProgress
+	}
+
+	bp.state = reconnecting
+	defer func() {
+		// Only reset to disconnected if we're still in reconnecting state
+		// (successful reconnection will set state to connected)
+		if bp.state == reconnecting {
+			bp.state = disconnected
+		}
+	}()
+
+	// Close existing connection if any
+	if bp.conn != nil {
+		_ = bp.conn.Close()
+		bp.conn = nil
+	}
+
+	// Gather local sequence numbers:
+	// - readerSeqNum: how many bytes we've successfully read from remote (send to remote)
+	// - writerSeqNum: how many bytes we've written so far (used for validation)
+	readerSeqNum := bp.reader.SequenceNum()
+	writerSeqNum := bp.writer.SequenceNum()
+
+	// Increment the generation and update both reader and writer.
+	// We do it now to track even the connections that fail during
+	// Reconnect.
+	bp.connGen++
+	bp.reader.SetGeneration(bp.connGen)
+	bp.writer.SetGeneration(bp.connGen)
+
+	// Send our reader sequence number to the remote and receive its reader sequence
+	// number. We will replay our outbound data from that point and backed pipe on the other
+	// side will replay from our reader sequence number.
+	conn, remoteReaderSeqNum, err := bp.reconnector.Reconnect(bp.ctx, readerSeqNum)
+	if err != nil {
+		return ErrReconnectFailed
+	}
+
+	// Validate sequence numbers
+	if remoteReaderSeqNum > writerSeqNum {
```

Comment:
The writer's `Reconnect()` method already handles this check, and it does so while holding the lock. Doing it here means we are subject to data races which could give false errors.